### PR TITLE
RN-675: Added 'External Database Connections' to Tupaia

### PR DESCRIPTION
**Pull request by rohan-bes**
### Issue RN-675:

Added a new concept, the 'External Database Connection' to Tupaia. These are required by SQL data tables, to control what database to pull from.

Fields:
- code: unique identifier code
- name: display name
- description: more detailed explanation of the connection
- permission_groups: which users can use this connection when creating SQL data tables

### CODE REVIEW

File: packages/database/src/modelClasses/ExternalDatabaseConnection.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    const establishedConnection = this.activeConnections[connectionId];
+    if (establishedConnection) {
+      return establishedConnection;
+    }
+
+    const host = requireEnv(`${EXT_DB_CONNECTION_ENV_VAR_PREFIX}_${code}_HOST`);
+    const port = getEnvVarOrDefault(`${EXT_DB_CONNECTION_ENV_VAR_PREFIX}_${code}_PORT`, 5432);
+    const database = requireEnv(`${EXT_DB_CONNECTION_ENV_VAR_PREFIX}_${code}_DATABASE`);
+    const user = requireEnv(`${EXT_DB_CONNECTION_ENV_VAR_PREFIX}_${code}_USER`);
+    const password = requireEnv(`${EXT_DB_CONNECTION_ENV_VAR_PREFIX}_${code}_PASSWORD`);
```
Comment (rohan-bes) [inline packages/database/src/modelClasses/ExternalDatabaseConnection.js:67]: To create a new connection, a user must first setup the environment variables and restart the server. Then they can create a record in the Admin Panel for the environment variables.

This allows us to:
a) Not have to store passwords in the database
b) Integrate easily with our CI/CD process

File: packages/central-server/src/apiV2/index.js [hunk 2]
```diff
@@ -229,6 +230,11 @@ apiV2.get('/dhisInstances/:recordId?', useRouteHandler(BESAdminGETHandler));
 apiV2.get('/dataServiceSyncGroups/:recordId?', useRouteHandler(GETSyncGroups));
 apiV2.get('/dataServiceSyncGroups/:recordId/logs', useRouteHandler(GETSyncGroupLogs));
 apiV2.get('/dataServiceSyncGroups/:recordId/logs/count', useRouteHandler(GETSyncGroupLogsCount));
+apiV2.get('/externalDatabaseConnections/:recordId?', useRouteHandler(BESAdminGETHandler));
```
Comment (rohan-bes) [inline packages/central-server/src/apiV2/index.js:233]: Only `BES_Admin` users can CRUD external database connections

File: packages/admin-panel/src/routes.js [hunk 3]
```diff
@@ -200,4 +201,16 @@ export const ROUTES = [
       },
     ],
   },
+  {
+    label: 'External Data',
```
Comment (rohan-bes) [inline packages/admin-panel/src/routes.js:205]: It will be nice to add DHIS2 Instances to this tab once we've got them in the Admin Panel @IgorNadj

Comment (IgorNadj) [reply packages/admin-panel/src/routes.js:205]: Yeah I was waiting until we consolidated the concept of data services before adding them to the admin panel. Right now it's kind of messy, we would have a tab for each type. Maybe we'll pop these db connections in as data services at that point too.

**Review by billli0 — APPROVED** [review]
Looks great! Just have a small change request on the sub url for consistence.

File: packages/admin-panel/src/routes.js [hunk 4]
```diff
@@ -200,4 +201,16 @@ export const ROUTES = [
...
+    to: '/externalDatabaseConnections',
```
[the remark this answers is not in the document]
Comment (rohan-bes) [reply packages/admin-panel/src/routes.js]: Ah good call! Thanks have fixed this 👍

**Pull request merged**